### ISSUE TITLE
Overlapping overloads false positive

### ISSUE THREAD

**Issue by jonathandung** [python/mypy#21353]
Two overloads involving `Self` and `Literal` that clearly do not overlap are flagged as overlapping.

**To Reproduce**

[Playground](https://mypy-play.net/?mypy=latest&python=3.12&gist=533aab86132dd720f644978a77187fe0)

**Expected Behavior**

Neither overload includes the other, so there should be no error.

**Actual Behavior**

main.py:4: error: Overloaded function signatures 1 and 2 overlap with incompatible return types  [overload-overlap]
Found 1 error in 1 file (checked 1 source file)

**Your Environment**

- Mypy version used: 1.20.2
- Python version used: 3.12

**Comment by ilevkivskyi** [comment]
You _clearly_ don't understand what overlapping meas, because it is almost a "textbook example" of unsafe overload overlap. Continuing your example:
```python
def test(x: object) -> None:
    assert not foo().bar(x)

test(foo())  # this will crash because the first overload will be selected at runtime.
```
